sql/src/main/java/io/crate/jobs/transport/NodeDisconnectJobMonitorService.java

```diff
@@ -83,16 +92,22 @@ public void onNodeConnected(DiscoveryNode node) {
 
     @Override
     public void onNodeDisconnected(final DiscoveryNode node) {
-        final Collection<JobExecutionContext> contexts = jobContextService.getContextsByCoordinatorNode(node.getId());
+        final Collection<UUID> contexts = jobContextService.getJobIdsByCoordinatorNode(node.getId()).collect(Collectors.toList());
         if (contexts.isEmpty()) {
-            return;
+            // Disconnected node is not a handler node --> kill jobs on all participated nodes
+            contexts.addAll(jobContextService.getJobIdsByParticipatingNodes(node.getId()).collect(Collectors.toList()));
+            KillJobsRequest killJobsRequest = new KillJobsRequest(contexts);
+            killJobsNodeAction.broadcast(killJobsRequest, new ActionListener<KillResponse>() {
+                @Override
+                public void onResponse(KillResponse killResponse) { }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    LOGGER.warn("failed to send kill request to nodes");
+                }
+            }, Arrays.asList(node.getId()));
```

Comment:
Probably because it takes a bit for the kill to propagate back to the handler node. But I guess it's fine to kill directly on the handler as well.